### ISSUE TITLE
Text search is always case sensitive for Non-ASCII letters if layer stored in SQLite (GPKG, SpatiaLite, FDO SQLite Spatial) 

### ISSUE THREAD

**Issue by phidrho** [qgis/QGIS#50863]
### What is the bug or the crash?

When I tried to search for attribute inside GPKG layer which contains letters used in Croatian alphabet (š,đ,č,ć,ž) via `Locator` with use of `af` modulator - I did not get any results if entered different case (uppercase or lowercase) than what is stored,
I tried to do the same thing for SHP layer and it works as expected (it's not case sensitive).

I also noticed small UI bug - in form for Locator Configuration (Setting>Options>Locator>Configuration) there is a widget (QSpinBox) to set up `Maximum number of results` which accepts values up to 200, but value cannot be above 99.

Both bugs are shown in [video demo at youtube](https://youtu.be/qNTo7e4SOKA).

### Steps to reproduce the issue

Download a sample project with GPKG and SHP variant of same layer - [qgis_locator.zip](https://github.com/qgis/QGIS/files/9985170/qgis_locator.zip)

1. Use a `Locator` with `af` prefix with any of Croatian letters - š, đ, č, ć, ž, Š, Đ, Č, Ć, Ž

2. for UI widget bug see end of video noted above

### Versions

QGIS version
3.28.0-Firenze
QGIS code revision
ed3ad0430f3
Qt version
5.15.3
Python version
3.9.5
Compiled against GDAL/OGR
3.5.2
Running against GDAL/OGR
3.5.3
PROJ version
9.1.0
EPSG Registry database version
v10.074 (2022-08-01)
GEOS version
3.10.3-CAPI-1.16.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
actions_for_relations
1.2.0
document_management_system
0.3.1
EquirectangularViewer
1.4
GroupStats
2.2.6
plugin_reloader
0.9.2
QGIS3-getWKT
1.4
QuickWKT
3.1
SelectByRelationship
0.3.3
db_manager
0.1.20
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

**Comment by agiudiceandrea** [comment]
The second reported issue may be related to the PR https://github.com/qgis/QGIS/pull/38628.

**Comment by krzysztof-adamski** [comment]
I think the problem with searching is not even Croatian alphabet, I have to same with Polish in filtering (GPKG layer)

Example
`
        query = " column_name in ('światłowodowe') "
        f_request = QgsFeatureRequest().setFilterExpression(query)
        features = layer.getFeatures(f_request)
        ids = [feature.id() for feature in features]
`
Always is zero results, but in that column value "światłowodowe" exist.
This behavior regarding custom plugin, but if I use python console in QGIS, and add above line : # -*- coding: utf-8 -*-
searching work well.

Is it a bug or something else ?

I trying somthing like: query = f" ({query}) COLLATE Polish_CI_AS " - it wont work

**Comment by phidrho** [comment]
Hi @krzysztof-adamski,

yes, it's not only Croatian alphabet, I just used this for example, it's probably all letters outside [ASCII symbols table](https://theasciicode.com.ar/) as I noted in title.
Developers are aware of that, it's common problem, and when bug gets fixed, it will fix all worldwide letters.

**Comment by krzysztof-adamski** [comment]
@phidrho I am not super duper in QGIS, sql, python etc, I figure out this problem just like that

    def replace_polish_letter(self, column_name):
        polish_chars = ["ą", "ó", "ę", "ł", "ś", "ć", "ż", "ź", "ń"]

        string = "upper(#COLUMN#)"
        prefix = suffix = ""
        for polish_char in polish_chars:
            prefix += f"replace("
            suffix += f", '{polish_char}', '{unidecode(polish_char)}')"

        concat = prefix + string + suffix
        concat = concat.replace("#COLUMN#", column_name)
```
query = query.replace(column_name, "#COLUMN#")
query = query.upper()
query = query.replace("#COLUMN#", self.replace_polish_letter(column_name))
```
I know it is weird, but enough for yet.

**Comment by sebastic** [comment]
FWIW sqlite3 in Debian is no longer built with ICU support, see: [#584506 - libsqlite3-0: Include Unicode support](https://bugs.debian.org/584506).

Related bugreport: [#536662 - sqlite3: CLI not UTF-8 clean](https://bugs.debian.org/536662)

**Comment by jgrocha** [comment]
Using GDAL compiled from master, it works in QGIS. It works in the command line too.

I have two features in QGIS:

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN"><html><head></head><body>
wkt_geom | fid | name
-- | -- | --
Point (-7.74873431926051559 37.09479248838015764) | 2 | Šimun
Point (-7.74289089247152074 37.08442040582968957) | 3 | Maša

</body></html>

Searching in QGIS works.

![image](https://github.com/user-attachments/assets/b37e28d8-fe87-4c05-b714-fb4db6efddad)


The following query also works on the command line:

```
ogr2ogr -sql "SELECT m."fid", m."name" FROM "test" m WHERE lower("name") LIKE lower('%Š%') ESCAPE '\'" -f CSV /vsistdout/ bratislava.gpkg 
name
Šimun
Maša
```

**Comment by jgrocha** [comment]
The GDAL query works, not because `lower` or `upper` works. **These functions are not working for diacritics.**

The GDAL SQL works because the `LIKE` operator does an insensitive case comparison.

This also works, without the `lower` function applied to both sides of the comparison:
```bash
ogr2ogr -sql "SELECT m."fid", m."name", lower(m."name"), upper(m."name") FROM "test" m WHERE "name" LIKE '%Š%' ESCAPE '\'" -f CSV /vsistdout/ bratislava.gpkg 
name,lower(m.name),upper(m.name)
Šimun,Šimun,ŠIMUN
Maša,maša,MAšA
```

**Comment by rouault** [comment]
> The GDAL SQL works because the `LIKE` operator does an insensitive case comparison.

and because since GDAL 3.9, GDAL overloads the base LIKE operator to support case-insentitive UTF-8 comparisons, but restricted to a few character-set (mostly European ones): https://gdal.org/en/latest/user/sql_sqlite_dialect.html#like-operator

**Comment by phidrho** [comment]
Hi @rouault,

I also tested now on my local machine and can confirm it works.
I compiled latest GDAL, and in ogr2ogr works as expected:

GDAL 3.4.1, released 2021/12/27 (Ubuntu 22.04 official repository)

```sh
$ ogr2ogr -sql "SELECT m."fid", m."name" FROM "test" m WHERE lower("name") LIKE lower('%Š%') ESCAPE '\'" -f CSV /vsistdout/ /home/phidrho/Desktop/test.gpkg
GPKG: GeoPackage v1.2.0
GDAL: GDALOpen(/home/phidrho/Desktop/test.gpkg, this=0x5bb8bcb461a0) succeeds as GPKG.
GDAL: GDALDriver::Create(CSV,/vsistdout/,0,0,0,Unknown,(nil))
name,
Šimun
GPKG: finalize 0x5bb8bcb4c638
GDALVectorTranslate: 1 features written in layer 'SELECT'
GDAL: GDALClose(/home/phidrho/Desktop/test.gpkg, this=0x5bb8bcb461a0)
GDAL: GDALClose(/vsistdout/, this=0x5bb8bce69d70)
GDAL: In GDALDestroy - unloading GDAL shared library.
```

GDAL 3.10.0dev-25eb9f6ee7

```
$ LD_LIBRARY_PATH=. CPL_DEBUG=ON ./apps/ogr2ogr -sql "SELECT m."fid", m."name" FROM "test" m WHERE lower("name") LIKE lower('%Š%') ESCAPE '\'" -f CSV /vsistdout/ /home/phidrho/Desktop/test.gpkg
GPKG: GeoPackage v1.2.0
GDAL: GDALOpen(/home/phidrho/Desktop/test.gpkg, this=0x571e68570ff0) succeeds as GPKG.
GDAL: GDALDriver::Create(CSV,/vsistdout/,0,0,0,Unknown,(nil))
OGR2OGR: Using WriteArrowBatch()
GPKG: finalize 0x571e68570558
name
Maša
Šimun
GDAL: GDALClose(/home/phidrho/Desktop/test.gpkg, this=0x571e68570ff0)
GDAL: GDALClose(/vsistdout/, this=0x571e685b8ce0)
GDAL: In GDALDestroy - unloading GDAL shared library.
```

I would like to close this bug report, as it is a matter of time for fix to be applied, and it depends on packaging, not anymore on development.

I'm only notifying [packaging team](https://qgis.org/community/organisation/#packaging-team): @jef-n, @kyngchaos, @PeterPetrik, @kalxas, @daniviga, @sebastic
